### PR TITLE
files related to engagement semester

### PR DESCRIPTION
**Pull request by zfred219**
- Search app by semester regardless of format
- Search app by email
- Add engagement with semester field
- Show number of projects after applying filter
- Fix toggle checkbox

### CODE REVIEW

**Comment by EnjiaXu** [comment]
Some files here are already in the previous PRs (need to include them to pass CI).
We will comment on which files to ignore and which ones that relate to the different features.

File: Gemfile [hunk 1]
```diff
@@ -18,7 +19,6 @@ gem 'jbuilder', '~> 2.0'
 gem 'bootstrap_form'
 gem 'paperclip', '~> 5.2.1'
 gem 'aws-sdk', '~> 2.3.0'
-
 group :development do
   # Access an IRB console on exception pages or by using <%= console %> in views
   gem 'web-console', '~> 2.0'
```
Comment (EnjiaXu) [inline Gemfile:24]: Ignore this file (Included in a previous PR).

File: Gemfile.lock [hunk 2]
```diff
@@ -372,6 +374,7 @@ DEPENDENCIES
   jquery-rails
   json
   launchy
+  mimemagic (~> 0.3.10)
   omniauth-github
   paperclip (~> 5.2.1)
   pg (< 1.0.0)
```
Comment (EnjiaXu) [inline Gemfile.lock:380]: Ignore this file (Included in a previous PR).

File: app/views/apps/show.html.haml [hunk 3]
```diff
@@ -12,7 +12,8 @@
 .page-header
   %h1
     = link_to_unless @app.deployment_url.blank?, @app.name, @app.deployment_url, :target => '_blank'
-    = link_to_unless @app.repository_url.blank?, image_tag('octocat.png'), @app.repository_url, :target => '_blank'
+    = link_to_unless @app.repository_url.blank?, image_tag('octocat.png', title: 'GitHub Repository'), @app.repository_url, :target => '_blank'
+    = link_to_unless @app.pivotal_tracker_url.blank?, image_tag('tracker.png', size: '45x45', title: 'Pivotal Tracker'), @app.pivotal_tracker_url, :target => '_blank'
     - unless (url = @app.code_climate_url).blank?
       = link_to image_tag(url + '/badges/gpa.svg'), url, :target => '_blank'
       = link_to image_tag(url + '/badges/coverage.svg'), url, :target => '_blank'
```
Comment (EnjiaXu) [inline app/views/apps/show.html.haml:19]: Ignore this file (already in a previous PR)

File: app/views/creation/new.html.haml [hunk 4]
```diff
@@ -12,7 +12,7 @@
                 = user_form.text_field :name, label: 'User Name'
                 = user_form.email_field :email, label: 'Email'
                 = user_form.phone_field :preferred_contact, label: 'Preferred Contact'
-                = user_form.text_field :github_uid, label: 'Github Uid'
+                = user_form.text_field :github_uid, label: 'GitHub Username'
                 = user_form.collection_select :user_type, User.user_types, :first, ->(x){x.first.humanize}, {label: 'User Type'}, {class: 'select2'}
                 = user_form.text_field :sid, label: 'SID'
         %fieldset.col-sm-4
```
Comment (EnjiaXu) [inline app/views/creation/new.html.haml:18]: Ignore this file (already in a previous PR)

File: app/views/engagements/_index.html.haml [hunk 5]
```diff
@@ -31,4 +33,4 @@
             = link_to('Edit', edit_app_engagement_path(app,engagement), :class => 'btn btn-primary', style: 'padding: 5px 5px;')
             = link_to 'Destroy', app_engagement_path(app,engagement), method: :delete, data: { confirm: 'Are you sure?' }, :class => 'btn btn-danger', style: 'padding: 5px 5px;'
             = link_to('CSV', export_path(app,engagement), :class => 'btn btn-info', style: 'padding: 5px 5px;')
-            = link_to('Request Feedback', new_engagement_iteration_path(engagement), :class => 'btn btn-success pull-right',  style: 'padding: 5px 10px;')
\ No newline at end of file
+            = link_to('Request Feedback', new_engagement_iteration_path(engagement), :class => 'btn btn-success pull-right',  style: 'padding: 5px 10px;')
```
Comment (EnjiaXu) [inline app/views/engagements/_index.html.haml:36]: Feature: add semester info to engagements

File: app/views/layouts/_nav_header.html.haml [hunk 6]
```diff
@@ -1,32 +1,49 @@
...
-            = check_box_tag category, 1, (if @filters.nil? then true else @filters.include?(category) end), :class => 'filter'
-            = label_tag category, category
+      - if not current_user&.student?
+        = link_to current_iteration_path, :class => "text-light text-light-nav px-1" do
+          Current iteration
+      - if current_user&.coach? || current_user&.student?
+        = link_to matching_path, :class => "text-light text-light-nav px-1" do
+          Project Matching
+          %i.fas.fa-random
+      = link_to "Logout @#{current_user.github_uid}", logout_path, :class => "text-light text-light-nav px-1"
```
Comment (EnjiaXu) [inline app/views/layouts/_nav_header.html.haml:49]: Ignore this file (already in a previous PR)

File: app/controllers/search_controller.rb [hunk 7]
```diff
@@ -3,7 +3,7 @@ class SearchController < ApplicationController
 
   def search
```
Comment (zfred219) [inline app/controllers/search_controller.rb:4]: Add semester and email search functionalities in search controller

File: app/views/my_projects/index.html.haml [hunk 8]
```diff
@@ -61,19 +90,20 @@
...
             %br
-        %td{style: 'white-space: nowrap'}
-          - if current_user.coach? || current_user.client?
-            = link_to ApplicationHelper.edit_request_button_text(app), new_my_project_edit_path(app_id: app.id), :class => 'btn btn-primary'
+            - app.engagements.each do |engagement|
+              %span #{engagement.id} (#{engagement.team_number}): #{engagement.student_names}
+              %br
+          %td{style: 'white-space: nowrap'}
+            - if current_user.coach? || current_user.client?
+              = link_to ApplicationHelper.edit_request_button_text(app), new_my_project_edit_path(app_id: app.id), :class => 'btn btn-primary'
```
Comment (EnjiaXu) [inline app/views/my_projects/index.html.haml:109]: Feature: add pagination to project index page

File: app/views/users/_form.html.haml [hunk 9]
```diff
@@ -4,7 +4,7 @@
   = f.text_field :name, label: 'User Name'
   = f.email_field :email, label: 'Email'
   = f.phone_field :preferred_contact, label: 'Preferred Contact'
-  = f.text_field :github_uid, label: 'Github Uid'
+  = f.text_field :github_uid, label: 'GitHub Username'
   = f.collection_select :user_type, User.user_types, :first, ->(x) { x.first.humanize }, {label: 'User Type'}, {class: 'select2'}
   = f.text_field :sid, label: 'SID'
   = f.file_field :profile_picture, label: 'Profile Picture'
```
Comment (EnjiaXu) [inline app/views/users/_form.html.haml:10]: Ignore this file (already in a previous PR)

File: app/views/layouts/application.html.haml [hunk 10]
```diff
@@ -10,11 +14,26 @@
       <script src="https://cdnjs.cloudflare.com/ajax/libs/axe-core/3.4.0/axe.min.js" integrity="sha256-3uvAwJcMTsYRtepeqr32rrpy5zZ/fMYjiKEKPZM6Rtc=" crossorigin="anonymous"></script>
 
   %body
+    = render 'layouts/nav_header'
+    .sidenav{role: 'navigation', title: 'filters'}
+      = yield :filters
+        
     #main.container{role: 'main'}
-      = render 'layouts/nav_header'
+      = render 'layouts/nav_header_search'
```
Comment (zfred219) [inline app/views/layouts/application.html.haml:22]: **Search**
- Add search bar to layout

File: app/views/users/index.html.haml [hunk 11]
```diff
@@ -42,3 +47,33 @@
...
+      </form>
+    </div>`;
+    $('[data-toggle="popover"]').popover({
+      container: 'body',
+      html: true,
+      placement: 'bottom',
+      sanitize: false,
+      content: html
+    })
+  });
```
Comment (EnjiaXu) [inline app/views/users/index.html.haml:79]: Ignore this file (already in a previous PR)

File: app/views/users/show.html.haml [hunk 12]
```diff
@@ -32,4 +32,4 @@
   %h1 Engagements
   %span.help-block This user is participating in the following Engagement(s):
 
-= render partial: 'engagements/index', locals: {engagements: @user.participating_engagements}
\ No newline at end of file
+= render partial: 'engagements/index', locals: {engagements: @user.participating_engagements}
```
Comment (EnjiaXu) [inline app/views/users/show.html.haml:35]: Ignore this file (already in a previous PR)

File: features/search.feature [hunk 13]
```diff
@@ -6,10 +6,10 @@ Feature: searching to quickly find apps according to name/organization/descripti
 
 Background: Logged in
     Given the following apps exist:
-        | name   | description             | org_id | status  |
-        | app 1  | this is one test        | 1      | pending |
-        | app 2  | this is two test        | 2      | pending |
-        | app 3  | this is three test      | 3      | pending |
+        | name   | description             | org_id | status  | id |
```
Comment (zfred219) [inline features/search.feature:9]: **Search**
- Tests of search semesters

File: config/initializers/assets.rb [hunk 14]
```diff
@@ -9,3 +9,8 @@
 # Precompile additional assets.
 # application.js, application.css, and all non-JS/CSS in app/assets folder are already added.
 # Rails.application.config.assets.precompile += %w( search.js )
+# Rails.application.config.assets.precompile += %w( main_page.js )
+Rails.application.config.assets.precompile += %w( apps.scss )
+Rails.application.config.assets.precompile += %w( matching.css )
+Rails.application.config.assets.precompile += %w( progress.css )
+Rails.application.config.assets.precompile += %w( navbar.css )
```
Comment (EnjiaXu) [inline config/initializers/assets.rb:16]: Ignore this file (already in a previous PR)

File: config/routes.rb [hunk 15]
```diff
@@ -14,12 +31,20 @@
...
   resources :comments, :only => [:edit, :destroy]
+
+  get 'users/import' => 'users#import'
+
   resources :users do
     resources :comments, only: [:create, :update], module: :users
+    collection {post :import}
   end
 
   # my_projects routes
```
Comment (EnjiaXu) [inline config/routes.rb:50]: Ignore this file (already in a previous PR)

File: features/step_definitions/search_steps.rb [hunk 16]
```diff
@@ -19,6 +19,14 @@
   end
 end
 
+Given (/^the following engagements exist for "([^"]*)":$/) do |app_name, table|
```
Comment (zfred219) [inline features/step_definitions/search_steps.rb:22]: **Search**
- Allow to create mock data for search tests

File: db/migrate/20210303105028_add_semester_to_iterations.rb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+class AddSemesterToIterations < ActiveRecord::Migration
+  def change
+    add_column :engagements, :semester, :string
+  end
+end
```
Comment (EnjiaXu) [inline db/migrate/20210303105028_add_semester_to_iterations.rb:5]: Feature: add semester info to engagements

File: features/app_features.feature [hunk 18]
```diff
@@ -60,6 +60,27 @@ Scenario: A user can add initial features while editing apps
...
+
+# Testing pivotal tracker url in app edit
+Scenario: A user can add pivotal tracker url while editing apps
+   Given I follow "app1"
+   And I follow "Edit App"
+   When I fill in "Pivotal Tracker Url" with "https://www.pivotaltracker.com/12345"
+   And I press "Update App"
+   And I follow "app1"
+   Then I should see "Pivotal Tracker"
+   And I should see "https://www.pivotaltracker.com/12345"
```
Comment (EnjiaXu) [inline features/app_features.feature:86]: Ignore this file (already in a previous PR)

File: app/controllers/apps_controller.rb [hunk 19]
```diff
@@ -12,7 +11,68 @@ def index
...
+    @filtered_count = 0
+    App.find_each do |app|
+      if @deployment_statuses.key? app.status.to_sym
+        @filtered_count += 1
+      end
+    end
+
+    ## limit first several + offset to count start
+    offset = @each_page*(@page_num-1) < 0 ? 0 : @each_page*(@page_num-1)
+    @apps = App.limit(@each_page).offset(offset)
```
Comment (zfred219) [inline app/controllers/apps_controller.rb:75]: **Show number of projects after applying filter**

File: features/app_json.feature [hunk 20]
```diff
@@ -15,36 +15,37 @@ Background: apps with different status exist
...
+#			  "repository_url": "test3@test.com",
+#			  "most_recent_screencast_url": "",
+#			  "most_recent_screenshot_url": "",
+#			  "org": {
+#				"name": "org1",
+#				"url": null
+#			  }
+#			}
+#		]
+#	"""
```
Comment (EnjiaXu) [inline features/app_json.feature:51]: Feature: restrict user access based on their user type (student/client/coach)

File: app/controllers/my_projects_controller.rb [hunk 21]
```diff
@@ -8,9 +8,69 @@ def index
...
+    
+        @filtered_count = 0
+        App.unscoped.where(:org_id => orgs).find_each do |app|
+          if @deployment_statuses.key? app.status.to_sym
+            @filtered_count += 1
+          end
+        end
+
+        offset = @each_page*(@page_num-1) < 0 ? 0 : @each_page*(@page_num-1)
+        @apps = App.for_orgs(orgs, limit=@each_page, offset=offset)
```
Comment (zfred219) [inline app/controllers/my_projects_controller.rb:73]: **Show number of projects after applying filter**
- For My project view/page

File: features/clear_form.feature [hunk 22]
```diff
@@ -187,12 +187,12 @@ Scenario: user can reset a form for Edit User
...
+	And I fill in "GitHub Username" with "afox"
 	And I select "Coach" from "User Type"
 	And I press "Reset"
 	Then I should have filled in "Armando Fox" for "User Name"
 	And I should have filled in "armandofox@berkeley.edu" for "Email"
-	And I should have filled in "armandofox" for "Github Uid"
+	And I should have filled in "armandofox" for "GitHub Username"
 	And the "User Type" field should contain "coach"
 
 # Story ID: 153070009
```
Comment (EnjiaXu) [inline features/clear_form.feature:198]: Ignore this file (already in a previous PR)

File: features/create.feature [hunk 23]
```diff
@@ -57,7 +57,7 @@ Scenario: User can submit successfully if form is complete
         | User Name           | Fakeuser              |
         | Email               | fakeuser@berkeley.edu |
         | Preferred Contact   | 555-555-5555          |
-        | Github Uid          | fakegithubuid         |
+        | GitHub Username     | fakegithubuid         |
     And I fill in the "Org Information" fields as follows:
         | field                     | value                  |
         | Organization Name         | Group 20               |
```
Comment (EnjiaXu) [inline features/create.feature:63]: Ignore this file (already in a previous PR)

File: features/create_user.feature [hunk 24]
```diff
@@ -68,5 +68,3 @@ Scenario: When I click back I go back to the users page
     Given I am on the new user page
     And I follow "Back"
     Then I am on the users page
```
Comment (EnjiaXu) [inline features/create_user.feature:70]: Ignore this file (already in a previous PR)

File: features/edit_user.feature [hunk 25]
```diff
@@ -54,5 +54,3 @@ Scenario: I cannot submit with user name field blank
 Scenario: When I click back I go back to the users page
     And I follow "Back"
     Then I am on the users page
```
Comment (EnjiaXu) [inline features/edit_user.feature:56]: Ignore this file (already in a previous PR)

File: features/engagement_sem.feature [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+   And I follow "Apps"
+
+# Story ID: 176932021
+Scenario: A user can see the semester for each iteration on App's detail page
+Given I follow "app1"
+Then I should see "Semester"
+And I should see "FA17"
+And I should see "SP18"
+And I should see "SU18"
+
```
Comment (EnjiaXu) [inline features/engagement_sem.feature:47]: Feature: add semester info to engagements

File: app/views/apps/index.html.haml [hunk 27]
```diff
@@ -1,55 +1,84 @@
...
+      %div#deploy
+        %strong Deployment Status:
+        %em #{" (" + @total_deploy.to_s + ")"}
         %br
-    
-  %label.checkbox-inline
-    %div#vetting
-      %strong Vetting Status:
-      %em #{" (" + @total_vet.to_s + ")"}
+        .checkbox-inline
```
Comment (zfred219) [inline app/views/apps/index.html.haml:21]: **Fix checkbox toggle**

File: features/myprojects_pagination.feature [hunk 28]
```diff
@@ -71,7 +71,7 @@ Scenario: A client with many apps should see the first 10 apps by default
   Then I should see 100 apps starting with app 1 with sequence format:
     | name  | description     |
     | app   | app-description |
-  When I follow "Next"
+  When I follow "2"
   Then I should see 100 apps starting with app 101 with sequence format:
     | name  | description     |
     | app   | app-description |
```
Comment (EnjiaXu) [inline features/myprojects_pagination.feature:77]: Feature: add pagination to project index page

File: features/pagination.feature [hunk 29]
```diff
@@ -136,21 +139,22 @@ Scenario: I can click "First" or "Last" to see the first page or last page of ap
...
+#    Given I am on the apps page
+#    When I follow "All"
+#    And I follow "Next"
+#    Then I should see "app_k"
+#    When I follow "Next"
+#    Then I should see "app_k"
+#    And I should see "You are already on the LAST page."
 
 Scenario: There are some click buttons on orgs page that I can control the pagination
     Given I am on the orgs page
```
Comment (EnjiaXu) [inline features/pagination.feature:160]: Feature: add pagination to project index page

File: features/public_search.feature [hunk 30]
```diff
@@ -24,13 +24,13 @@ Background: Logged Out
...
-Scenario: search for an app by name keyword
-    Given I publicly search for "1"
-    Then I should see "app 1"
-    But I should not see "app 2"
-    And I should not see "app 3"
+#Scenario: search for an app by name keyword
+   # Given I publicly search for "1"
+    #Then I should see "app 1"
+   # But I should not see "app 2"
+    #And I should not see "app 3"
```
Comment (EnjiaXu) [inline features/public_search.feature:36]: Ignore this file (already in a previous PR)

File: features/search.feature [hunk 31]
```diff
@@ -111,19 +135,83 @@ Scenario: search for an organization by keyword
...
+Scenario: search for engagement semesters only with fuzzy input semester and year
+    Given I uncheck "Apps"
+    And I uncheck "Organizations"
+    And I uncheck "Users"
+    And I search for "Fa15"
+    Then I should see "app 1"
+    Then I should see "app 2"
+    Then I should not see "app 3"
+    
+
```
Comment (EnjiaXu) [inline features/search.feature:217]: Feature: add semester info to engagements

File: features/step_definitions/auth_user_steps.rb [hunk 32]
```diff
@@ -1,3 +1,30 @@
...
+                              :uid => '6789',
+                              :info => {
+                                  :name => user['name'],
+                                  :nickname => user['github_uid'],
+                                  :email => user['email'],
+                                  :type_user => user["staff"]
+                              }
+                            }
+                          )
+end
```
Comment (EnjiaXu) [inline features/step_definitions/auth_user_steps.rb:30]: Feature: restrict user access based on usertype (student/client/coach)

File: features/support/env.rb [hunk 33]
```diff
@@ -52,9 +52,14 @@
...
+options = {
+  :window_size => [6000, 3000]
+}
 # Capybara driver
 require 'capybara/poltergeist'
+Capybara.register_driver :poltergeist do |app|
+   Capybara::Poltergeist::Driver.new(app, options)
+end
 Capybara.javascript_driver = :poltergeist
 Capybara.default_max_wait_time = 1
```
Comment (EnjiaXu) [inline features/support/env.rb:65]: Ignore this file (already in a previous PR)

File: features/type_users.feature [hunk 34]
```diff
@@ -143,4 +143,4 @@ Scenario: User can submit a create form that includes user type and SID for Staf
     | Code Climate Url   | Fake app codeclimate url |
   And I press "Submit"
   Then I should be on the app details page for "Fake app"
-  Then I should see "User, Org, and App were successfully created"
\ No newline at end of file
+  Then I should see "User, Org, and App were successfully created"
```
Comment (EnjiaXu) [inline features/type_users.feature:146]: Ignore this file (already in a previous PR)

File: features/user_coach_auth.feature [hunk 35]
```diff
@@ -195,4 +167,18 @@ Scenario: Sad Path - Cannot create new iteration, edit and destroy existing iter
...
+  Then I should see "No user with GitHub name 'esaas_developer'."
+  And I should see "Login to ESaaS Engagements"
+  And I should see "Log in with GitHub"
+
+
+Scenario: Sad Path - If there github return callback error or potential Oauth call error
+  Given I login with github callback error
+  Then I should see "Authentication failed, please try again."
+  And I should see "Login to ESaaS Engagements" 
+  And I should see "Log in with GitHub"
```
Comment (EnjiaXu) [inline features/user_coach_auth.feature:184]: Feature: restrict user access based on usertype (student/client/coach)

File: features/user_view.feature [hunk 36]
```diff
@@ -48,6 +49,7 @@ Scenario: The User profile page should have the correct information
     And I should see "2017-11-02"
     And I should see "178"
     And I should see "Armando Fox"
+		And I should see "GitHub Username"
     And I should not see "ACElab"
     And I should not see "Teamscope"
-
+		And I should not see "Github ID"
```
Comment (EnjiaXu) [inline features/user_view.feature:55]: Ignore this file (already in a previous PR)

File: spec/controllers/iterations_controller_spec.rb [hunk 37]
```diff
@@ -3,6 +3,7 @@
 describe IterationsController, type: :controller do
   before do
     allow_any_instance_of(ApplicationController).to receive(:logged_in?).and_return(true)
+    allow_any_instance_of(ApplicationController).to receive(:check_student).and_return(true)
     @iteration = FactoryBot.create(:iteration)
     @engagement = FactoryBot.create(:engagement)
     @feedback_params = {
```
Comment (EnjiaXu) [inline spec/controllers/iterations_controller_spec.rb:9]: Feature: restrict user access based on usertype (student/client/coach)

File: spec/controllers/users_controller_spec.rb [hunk 38]
```diff
@@ -3,6 +3,7 @@
 describe UsersController, type: :controller do
   before do
     allow_any_instance_of(ApplicationController).to receive(:logged_in?).and_return(true)
+    allow_any_instance_of(ApplicationController).to receive(:check_student).and_return(true)
     @user = FactoryBot.create(:user)
     @user_params = {
         :name => "user1",
```
Comment (EnjiaXu) [inline spec/controllers/users_controller_spec.rb:9]: restrict user access based on usertype (student/client/coach)

File: spec/models/engagement_spec.rb [hunk 39]
```diff
@@ -86,5 +87,40 @@
...
+            eng.start_date = DateTime.new(2017,12,31,8)
+            expect(eng.get_semester()).to eq('FA17')
+
+            eng.start_date += 1.days
+            expect(eng.get_semester()).to eq('SP18')
+        end
+        
+    end
 
 end
```
Comment (EnjiaXu) [inline spec/models/engagement_spec.rb:126]: Feature: add semester info to engagements

File: features/vetting_search_dropdown.feature [hunk 40]
```diff
@@ -37,10 +38,10 @@ Scenario: I can search the dropdown list of status on new app page (including ve
...
   Given I am not logged in
-  And I am on the apps page
-  And I follow "New App"
   And I am on the login page
   And I follow "Log in with GitHub"
+  And I am on the apps page
+  And I follow "New App"
   And I am on the new_app page
   And I press "Create App"
   # id for original (deployment) status
```
Comment (EnjiaXu) [inline features/vetting_search_dropdown.feature:47]: Feature: restrict user access based on usertype (student/client/coach)

File: app/assets/stylesheets/apps.scss [hunk 41]
```diff
@@ -0,0 +1,16 @@
...
+    padding-top: 60px;
+    padding-left: 2rem;
+    margin-top: 0px;
+}
+  
+.container {
+    margin-left: 195px; /* Same as the width of the sidebar */
+    
+}
+
```
Comment (EnjiaXu) [inline app/assets/stylesheets/apps.scss:16]: Ignore this file (already in a previous PR)

File: app/assets/stylesheets/navbar.scss [hunk 42]
```diff
@@ -0,0 +1,37 @@
...
+    height: 200;
+    border: none;
+    border: 1px solid rgb(36, 7, 7);
+    border-radius: 0.5rem;
+    padding: 0.3rem;
+}
+
+.search-btn:hover {
+    border: 1px solid rgb(255, 255, 255);
+}
```
Comment (EnjiaXu) [inline app/assets/stylesheets/navbar.scss:37]: Ignore this file (already in a previous PR)

File: app/assets/stylesheets/site.scss [hunk 43]
```diff
@@ -86,3 +86,9 @@ a { color: #2b4988}
   padding-bottom: 5px;
   padding-left: 3px;
 }
+
+.required::after {
+    content: " *";
+    color: red;
+}
+
```
Comment (EnjiaXu) [inline app/assets/stylesheets/site.scss:94]: Ignore this file (already in a previous PR)

File: app/controllers/application_controller.rb [hunk 44]
```diff
@@ -57,14 +58,14 @@ def page_default_and_update(name, total_item)
...
+    @page_num = {"Previous"=>page_num-1,"Next"=>page_num+1,"First"=>1,"Last"=>@max_page_num, nil => page_num}[params["#{name}_page_action"]].to_i
     flash.now[:alert] = "You are already on the FIRST page." if @page_num == 0
-    flash.now[:alert] = "You are already on the LAST page." if @page_num == max_page_num + 1
-    @page_num = [[1,@page_num].max,max_page_num].min
+    flash.now[:alert] = "You are already on the LAST page." if @page_num == @max_page_num + 1
+    @page_num = [[1,@page_num].max,@max_page_num].min
     session["#{name}_page_num"] = @page_num.to_s
   end
-  
+
```
Comment (EnjiaXu) [inline app/controllers/application_controller.rb:68]: Feature: add pagination to project index page

File: app/controllers/application_controller.rb [hunk 45]
```diff
@@ -97,4 +98,41 @@ def get_reviewed_apps
...
+  def check_student
+    @@name_path = request.env['PATH_INFO']
+    if current_user&.student? and (@@name_path != "/my_projects" and @@name_path != "/")
+      begin
+        redirect_to :back, alert: "You do not have access to that page."
+      rescue Exception
+        redirect_to "/", alert: "You do not have access to that page."
+      end
+    end
+  end
```
Comment (EnjiaXu) [inline app/controllers/application_controller.rb:112]: Feature: restrict user access based on usertype (student/client/coach)

File: app/controllers/application_controller.rb [hunk 46]
```diff
@@ -97,4 +98,41 @@ def get_reviewed_apps
...
+        rescue Exception
+          redirect_to "/", alert: "You do not have a matching in progress."
+        end
+      else
+        matching_id = Engagement.find(engagement_id).matching.id
+        redirect_to show_engagement_matching_path(matching_id: matching_id, engagement_id: engagement_id)
+      end
+    end
+  end
+end
```
Comment (EnjiaXu) [inline app/controllers/application_controller.rb:138]: Feature: restrict user access for matching feature

File: app/controllers/session_controller.rb [hunk 47]
```diff
@@ -1,5 +1,6 @@
 class SessionController < ApplicationController
   skip_before_action :logged_in?
+  skip_before_action :check_student
   skip_before_action :auth_user?
   skip_before_action :verify_authenticity_token
```
Comment (EnjiaXu) [inline app/controllers/session_controller.rb:6]: Feature: restrict user access based on usertype (student/client/coach)

File: app/models/engagement.rb [hunk 48]
```diff
@@ -20,4 +20,19 @@ def summarize_customer_rating
...
+    if start_date < sp_end
+      sem = "SP"
+    elsif start_date < su_end
+      sem = "SU"
+    else
+      sem = "FA"
+    end
+    start_date.strftime("#{sem}%y")
+  end
 end
```
Comment (EnjiaXu) [inline app/models/engagement.rb:38]: Feature: add semester info to engagements

File: app/views/apps/_form.html.haml [hunk 49]
```diff
@@ -6,6 +6,7 @@
   = f.url_field :deployment_url, label: 'Deployment Url'
   = f.url_field :repository_url, label: 'Repository Url'
   = f.url_field :code_climate_url, label: 'Code Climate Url'
+  = f.url_field :pivotal_tracker_url, label: 'Pivotal Tracker Url'
   = f.text_area :features, rows: '5', label: 'App Initial Features'
   = f.collection_select :org_id, Org.all, :id, :name, {label: 'Organization'}, {class: 'select2'}
   = f.collection_select :status, App.statuses, :first, ->(x) { x.first.humanize}, {label: 'Status'}, {class: 'select2'}
```
Comment (EnjiaXu) [inline app/views/apps/_form.html.haml:12]: Ignore this file (already in a previous PR)

File: app/views/layouts/_nav_header_search.haml [hunk 50]
```diff
@@ -0,0 +1,13 @@
...
+.navbar.ml-5
+  = form_tag search_path, :method => :post do
+    = label_tag "keyword", "label", :class => "label"
+    = text_field_tag "keyword", "", :size=>"70", :style=>"height:2.5em;", :placeholder => "Try 'Spring 15'", :class=>"search-input"
+    = button_tag :type=>"submit", :class => "btn search-btn", :style=> "background-color: #ffd15c;" do
+      Search 
+      %i.fas.fa-search 
+    - %w(Apps Organizations Users Semesters).each do |category|
+      = check_box_tag category, 1, (if @filters.nil? then true else @filters.include?(category) end), :class => 'filter ml-3'
+      = label_tag category, category
```
Comment (EnjiaXu) [inline app/views/layouts/_nav_header_search.haml:13]: Ignore this file (already in a previous PR)

File: features/apps_controller.feature [hunk 51]
```diff
@@ -126,12 +126,9 @@ Scenario: Login to github to falsely edit an existing app
...
-  And I am on the apps page
-  And I want to destroy "app1"
   And I am on the login page
   And I follow "Log in with GitHub"
-  Then I should see "Write a comment"
-  And I follow "Apps"
+  And I am on the apps page
   And I want to destroy "app1"
   Then I should see "App was successfully destroyed."
   And I am on the apps page
```
Comment (EnjiaXu) [inline features/apps_controller.feature:134]: Feature: restrict user access based on usertype (student/client/coach)

File: features/login_github_redirect.feature [hunk 52]
```diff
@@ -70,17 +70,17 @@ Scenario: login with Github from create page, after visiting orgs and users
...
 Scenario: Login to github to edit an existing app successfully
   #Story ID: #152298585
   Given I am not logged in
-  And I am on the apps page
-  And I want to edit "app1"
   And I am on the login page
   And I follow "Log in with GitHub"
+  And I am on the apps page
+  And I want to edit "app1"
   And I should see "Editing App"
```
Comment (EnjiaXu) [inline features/login_github_redirect.feature:86]: Feature: restrict user access based on usertype (student/client/coach)

**Pull request merged**